light-node/src/commands/start.rs

```diff
@@ -108,53 +94,102 @@ impl config::Override<LightNodeConfig> for StartCmd {
+            status_err!("no trusted or verified state in store for primary, please initialize with the `initialize` subcommand first");
+            std::process::exit(1);
+        }
     }
+    // TODO: this should do proper error handling, be gerneralized
+    // then moved to to the light-client crate.
+    fn make_instance(
+        &self,
+        light_config: &LightClientConfig,
+        io: ProdIo,
+        options: light_client::Options,
+    ) -> Instance {
+        let peer_id = light_config.peer_id;
+        let db_path = light_config.db_path.clone();
 
-    // check that the val hash matches
-    let vals = req.validator_set(height).await?;
+        let db = sled::open(db_path).unwrap_or_else(|e| {
+            status_err!("could not open database: {}", e);
+            std::process::exit(1);
+        });
 
-    if vals.hash() != vals_hash {
-        // TODO
-        panic!("vals hash dont match")
-    }
+        let light_store = SledStore::new(db);
 
-    let signed_header = req.signed_header(height).await?;
+        let state = State {
+            light_store: Box::new(light_store),
+            verification_trace: HashMap::new(),
+        };
 
-    // TODO: validate signed_header.commit() with the vals ...
+        let verifier = ProdVerifier::default();
+        let clock = SystemClock;
+        let scheduler = scheduler::basic_bisecting_schedule;
 
-    let next_vals = req.validator_set(height + 1).await?;
+        let light_client = LightClient::new(peer_id, options, clock, scheduler, verifier, io);
 
-    // TODO: check next_vals ...
+        Instance::new(light_client, state)
+    }
 
-    let trusted_state = &State::new(signed_header, next_vals);
+    fn start_rpc_server<H>(h: H)
+    where
+        H: Handle + Send + Sync + 'static,
+    {
+        let server = Server::new(h);
+        let laddr = app_config().rpc_config.listen_addr;
+        // TODO(liamsi): figure out how to handle the potential error on run
+        std::thread::spawn(move || rpc::run(server, &laddr.to_string()));
+    }
+}
 
-    store.add(trusted_state.to_owned())?;
+impl StartCmd {
+    fn construct_supervisor(&self) -> Supervisor {
+        // TODO(ismail): we need to verify the addr <-> peerId mappings somewhere!
+        let mut peer_map = HashMap::new();
+        for light_conf in &app_config().light_clients {
+            peer_map.insert(light_conf.peer_id, light_conf.address.clone());
+        }
+        let io = ProdIo::new(
+            peer_map.clone(),
+            Some(app_config().rpc_config.request_timeout),
+        );
+        let conf = app_config().deref().clone();
+        let options: light_client::Options = conf.into();
+
+        let mut peer_list: PeerListBuilder<Instance> = PeerList::builder();
+        for (i, light_conf) in app_config().light_clients.iter().enumerate() {
+            let instance = self.make_instance(light_conf, io.clone(), options);
+            if i == 0 {
```

Comment:
Should we perhaps make the initial primary/witness distinction more explicit by introducing a `primary` option and a `witnesses` array in the config?

Reply:
Probably a good idea. If we make this more explicit, the question though is what happens if the (e.g.) the primary doesn't reply? Does the light-node try the other nodes and makes the first one primary? Or do we just return and let the user figure it out? I'll leave this one open too and capture it in a followup issues, too.

Reply:
Yeah that's a good question, both in general and from a UX perspective. Let's discuss it in a follow-up!